browser/ui/ai_rewriter/ai_rewriter_dialog_delegate.cc

```diff
@@ -86,6 +97,137 @@ class AIRewriterDialogDelegate::DialogContentsObserver
+    CHECK(browser);
+
+    auto* host_widget = views::Widget::GetWidgetForNativeWindow(
+        browser->window()->GetNativeWindow());
+    CHECK(host_widget);
+
+    host_widget_ = host_widget->GetWeakPtr();
+
+    host_widget_observation_.Observe(host_widget);
+    dialog_widget_observation_.Observe(dialog_widget);
+    host_observation_.Observe(host);
+
+    UpdatePosition(target_contents->GetFocusedFrame());
+  }
+
+  DialogPositioner(const DialogPositioner&) = delete;
+  DialogPositioner& operator=(const DialogPositioner&) = delete;
+  ~DialogPositioner() override = default;
+
+  void OnPositionRequiresUpdate() override {
+    if (!target_contents_) {
+      return;
+    }
+
+    UpdatePosition(target_contents_->GetFocusedFrame());
+    UpdateFocusedBounds();
+  }
+
+  void OnHostDestroying() override {
+    host_widget_observation_.Reset();
+    host_observation_.Reset();
+    dialog_widget_observation_.Reset();
+  }
+
+  void OnWidgetBoundsChanged(views::Widget* widget,
+                             const gfx::Rect& new_bounds) override {
+    OnPositionRequiresUpdate();
+  }
+
+ private:
+  void UpdateFocusedBounds() {
+    if (!target_contents_) {
+      return;
+    }
+    auto* browser = chrome::FindBrowserWithTab(target_contents_.get());
+    if (!browser) {
+      return;
+    }
+
+    mojo::Remote<mojom::AIRewriterAgent> agent;
+    auto* frame = target_contents_->GetFocusedFrame();
+    frame->GetRemoteInterfaces()->GetInterface(
+        agent.BindNewPipeAndPassReceiver());
+    auto* raw_agent = agent.get();
+    raw_agent->GetFocusBounds(base::BindOnce(
+        [](mojo::Remote<mojom::AIRewriterAgent> agent,
+           base::WeakPtr<AIRewriterDialogDelegate::DialogPositioner> positioner,
+           content::WeakDocumentPtr document_pointer,
+           const gfx::RectF& focus_rect) {
+          if (!positioner) {
+            return;
+          }
+          positioner->last_bounds_ = focus_rect;
+          positioner->UpdatePosition(
+              document_pointer.AsRenderFrameHostIfValid());
+        },
+        std::move(agent), weak_ptr_factory_.GetWeakPtr(),
+        frame->GetWeakDocumentPtr()));
+  }
+
+  void UpdatePosition(content::RenderFrameHost* rfh) {
+    if (!rfh || !last_bounds_ || !dialog_widget_ || !host_widget_) {
+      return;
+    }
+
+    auto transformed = TransformFrameRectToView(rfh, last_bounds_.value());
+
+    auto dialog_bounds = dialog_widget_->GetWindowBoundsInScreen();
+    dialog_bounds.set_x(transformed.CenterPoint().x() -
+                        dialog_bounds.width() / 2);
+    dialog_bounds.set_y(transformed.bottom() + top_);
+
+    auto host_bounds = host_widget_->GetWindowBoundsInScreen();
+    host_bounds.set_origin(gfx::Point());
+
+    if (constrained_window::PlatformClipsChildrenToViewport() &&
+        !host_bounds.Contains(dialog_bounds)) {
+      dialog_bounds.AdjustToFit(host_bounds);
+    }
+
+    dialog_widget_->SetBounds(dialog_bounds);
```

Comment:
I'm curious if we need to update top_ with dialog_bounds.top() here. Also what would happen when scrolling.

Reply:
For now we're ignoring scrolling - the positioning is more best effort than anything. Most of the other dialogs in Chromium don't account for scrolling either (i.e. the autofill popups).

Additionally, as this is a modal the user can't scroll the content behind manually (though the page could).

Reply:
Aha thanks for the explanation.